pkg/storage/stores/shipper/compactor/index_set.go

```diff
@@ -0,0 +1,337 @@
+
+	err = is.compactedDB.Close()
+	if err != nil {
+		return err
+	}
+
+	is.compactedDB = destDB
+	is.uploadCompactedDB = true
+	is.removeSourceObjects = true
+	is.compactedDBRecreated = true
+	return nil
+}
+
+// isReady checks whether indexSet is ready for usage and whether there was an error during initialization.
+// All the external usages should check the readiness before using an instance of indexSet.
+func (is *indexSet) isReady() error {
+	select {
+	case <-is.ready:
+		return is.err
+	case <-is.ctx.Done():
+		return is.ctx.Err()
+	}
+}
+
+// writeBatch writes a batch to compactedDB
+func (is *indexSet) writeBatch(_ string, batch []indexEntry) error {
+	is.uploadCompactedDB = true
+	is.removeSourceObjects = true
+	return is.compactedDB.Batch(func(tx *bbolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists(bucketName)
+		if err != nil {
+			return err
+		}
+
+		for _, w := range batch {
+			err = b.Put(w.k, w.v)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+// runRetention runs the retention on index set
+func (is *indexSet) runRetention(tableMarker retention.TableMarker) error {
+	empty, modified, err := tableMarker.MarkForDelete(is.ctx, is.tableName, is.compactedDB)
+	if err != nil {
+		return err
+	}
+
+	if empty {
+		is.uploadCompactedDB = false
+		is.removeSourceObjects = true
+	} else if modified {
+		is.uploadCompactedDB = true
+		is.removeSourceObjects = true
+	}
+
+	return nil
+}
+
+// upload uploads the compacted db in compressed format.
+func (is *indexSet) upload() error {
+	compactedDBPath := is.compactedDB.Path()
+
+	// close the compactedDB to make sure all the writes are processed.
+	err := is.compactedDB.Close()
+	if err != nil {
+		return err
+	}
+
+	is.compactedDB = nil
+
+	fileNameFormat := "%s.gz"
+	if is.compactedDBRecreated {
+		fileNameFormat = "%s" + recreatedCompactedDBSuffix
+	}
+	fileName := fmt.Sprintf(fileNameFormat, shipper_util.BuildIndexFileName(is.tableName, uploaderName, fmt.Sprint(time.Now().Unix())))
+
+	return uploadFile(compactedDBPath, func(file io.ReadSeeker) error {
+		return is.baseIndexSet.PutFile(is.ctx, is.tableName, is.userID, fileName, file)
+	}, is.logger)
+}
+
+// removeFilesFromStorage deletes source objects from storage.
+func (is *indexSet) removeFilesFromStorage() error {
+	level.Info(is.logger).Log("msg", "removing source db files from storage", "count", len(is.sourceObjects))
+
+	for _, object := range is.sourceObjects {
+		err := is.baseIndexSet.DeleteFile(is.ctx, is.tableName, is.userID, object.Name)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// done takes care of file operations which includes:
+// - recreate the compacted db if required.
+// - upload the compacted db if required.
+// - remove the source objects from storage if required.
+func (is *indexSet) done() error {
+	if !is.uploadCompactedDB && !is.removeSourceObjects && mustRecreateCompactedDB(is.sourceObjects) {
```

Comment:
may be it was supposed to be `is.mustRecreateCompactedDB` ?

Reply:
The usage is correct here, I was supposed to drop the `mustRecreateCompactedDB` method on the index set.